flash: stm32f3xx: fixup byte write method

erwango opened the pull request:
stm32f3xx flash could only be programmed 16 bits at a time.
For one byte programming, stm32f3xx flash driver applies AND
operation with 0x00FF to the byte to be written. This results
in programming 0x00 on leading byte of the half word, which
won't be writable anymore.
This commit forces leading byte to 0xFF (erase value) so it
could be written on a later write operation.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

galak reviewed (commented):
Can we work up a test to try and validate various flash write boundary cases like this?

erwango committed:
flash: stm32f3xx: fixup byte write method

stm32f3xx flash could only be programmed 16 bits at a time.
For one byte programming, stm32f3xx flash driver applies AND
operation with 0x00FF to the byte to be written. This results
in programming 0x00 on leading byte of the half word, which
won't be writable anymore.
This commit forces leading byte to 0xFF (erase value) so it
could be written on a later write operation.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@galak: Should be possible but would require some info like allowed write_block_size, flash area where we could write (should not over write app code obviously) and there might be some varying contraints depending on the SoCs. All in all this seems not a quick shoot.
Meanwhile sample/driver/flash_shell allow to test manually various configurations
Thinking about it, I'd think it would also be interesting to merge f3 into "generic" stm32 driver (today supporting F4 and L4) > I'll log an issue on this.

erwango commented:
Testing patch showed that write byte is possible, but it is not possible to write unaligned byte.
Hence, I think would be better to make this clear and remove possibility of writing single bytes.
I propose to close this PR and add this requirement to issue #4197
